components/base/kinematicbase/ptgKinematics.go

```diff
@@ -247,6 +278,79 @@ func (ptgk *ptgBaseKinematics) ErrorState(ctx context.Context, plan [][]referenc
 	return spatialmath.PoseBetween(nominalPose, actualPIF), nil
 }
 
+func (ptgk *ptgBaseKinematics) trajectoryToArcsteps(traj []*tpspace.TrajNode) []arcStep {
+	finalSteps := []arcStep{}
+	timeStep := 0.
+	curDist := 0.
+	// Trajectory distance is either length in mm, or if linear distance is not increasing, number of degrees to rotate in place.
+	lastLinVel := r3.Vector{0, traj[0].LinVel * ptgk.linVelocityMMPerSecond, 0}
+	lastAngVel := r3.Vector{0, 0, traj[0].AngVel * ptgk.angVelocityDegsPerSecond}
+	nextStep := arcStep{
+		linVelMMps:      lastLinVel,
+		angVelDegps:     lastAngVel,
+		timestepSeconds: 0,
+	}
+	for _, trajPt := range traj {
+		nextLinVel := r3.Vector{0, trajPt.LinVel * ptgk.linVelocityMMPerSecond, 0}
+		nextAngVel := r3.Vector{0, 0, trajPt.AngVel * ptgk.angVelocityDegsPerSecond}
+		if nextStep.linVelMMps.Sub(nextLinVel).Norm2() > 1e-6 || nextStep.angVelDegps.Sub(nextAngVel).Norm2() > 1e-6 {
+			// Changed velocity, make a new step
+			nextStep.timestepSeconds = timeStep
+			finalSteps = append(finalSteps, nextStep)
+			nextStep = arcStep{
+				linVelMMps:      nextLinVel,
+				angVelDegps:     nextAngVel,
+				timestepSeconds: 0,
+			}
+			timeStep = 0.
+		}
+		distIncrement := trajPt.Dist - curDist
+		curDist += distIncrement
+		if nextStep.linVelMMps.Y != 0 {
+			timeStep += distIncrement / (math.Abs(nextStep.linVelMMps.Y))
+		} else if nextStep.angVelDegps.Z != 0 {
+			timeStep += distIncrement / (math.Abs(nextStep.angVelDegps.Z))
+		}
+	}
+	nextStep.timestepSeconds = timeStep
+	finalSteps = append(finalSteps, nextStep)
+	return finalSteps
```

Comment:
I'm having a hard time reviewing this through inspection.  Have you tested this code on hardware?